### ISSUE TITLE
some issue with tabs reverting their value?

### ISSUE THREAD

**Issue by chriddyp** [plotly/dash-core-components#331]
It's being reported in the community forum: https://community.plot.ly/t/problem-with-callbacks-in-different-tabs/13044/13?u=chriddyp

**Comment by chriddyp** [comment]
Yeah, that sounds plausible. This is the first component that has demonstrated an implicit link between children and parents, and so I wouldn't be surprised if there was some surprising behaviour, probably when we're doing a `createElement` here: 
https://github.com/plotly/dash-renderer/blob/b1cfc996563bd0b57f9487b15212dd5b782e5b3a/src/TreeContainer.js#L75-L79

I'd try to keep the `Tabs` stateless when it's connected to dash-renderer. That is, instead of lines like:
```
        this.setState({
            selected: value,
        });
        if (this.props.setProps) {
            this.props.setProps({value: value});
}
```
have:
```
        if (this.props.setProps) {
            this.props.setProps({value: value});
      } else {
        this.setState({
            selected: value,
        });
      }
```

That is, if it's connected to dash-renderer (`setProps` is defined), then let `dash-renderer` control `selected` value just through props. If it's not connected to `dash-renderer`, then just handle the state internally in the component.

**Comment by valentijnnieman** [comment]
@chriddyp Ah yeah, so that's the same problem as [this issue](https://github.com/plotly/dash-renderer/issues/84) right? I tried to come up with a better solution for that but haven't yet. 

That suggestion is interesting and probably a good idea, but in the example above there are no callbacks assigned to the Tabs, so won't that leave `setProps` empty meaning it will use `setState` anyway?

**Comment by chriddyp** [comment]
Yeah, good point. Could you be a little more precise on this statement?
> Then, if I click on a radio button, the Tabs render() is somehow called again, and it's value is updated to it's default (1, taken from the parent Tabs component) again.

Do you mean the `Tab` render is called again, taken from the parent `Tabs` component? There is only one parent `Tabs` instance here.

**Comment by valentijnnieman** [comment]
Sorry, that was confusing, let me rewrite that.

Then, if I click on a radio button, `Tabs.render()` is called again, and `state.value` is now 1 again (which is how it was set in the app: `dcc.Tabs(value=1 ...`). Neither the `constructor()` nor the `selectHandler()` methods are being called (the only two places that modify state).

**Comment by chriddyp** [comment]
>  - the only two places in the Tabs component source code where state.selected is being changed.

It's also being updated in `componentWillReceiveProps` right?
https://github.com/plotly/dash-core-components/blob/d004b29fa78643867658a273d3a4fab14a28ce0f/src/components/Tabs.react.js#L153-L160

**Comment by valentijnnieman** [comment]
Oh, yes, you're right.

**Comment by chriddyp** [comment]
> Only, it shouldn't - there are no callbacks being fired that involve Tabs.

Yeah, so you can't assume that `dash-renderer` will only re-render your component when new properties change. Your component needs to be idempotent: safe to re-render by the parent multiple times.

In this case it seems excessive, but imagine if a callback updated a different property of the Tabs like `style`: we'd need to call `componentWillReceiveProps` with that new `style` property and re-render the table.

**Comment by valentijnnieman** [comment]
Ah right, I remember now. I think my thinking was this: when the Tab's value get's updated, for example in `selectHandler`, it updates it in `this.state` and also fires `setProps()` to update the value in the Dash, so that Dash knows the `value` prop has been changed. But because there's no callback, I guess `setProps` isn't set, so it isn't updating in Dash, which, when it rerenders, it will rerender with what Dash thinks it's value should be. 

What's the solution then? If we want users to use the Tabs without a callback, then I don't see a way of updating the value prop in Dash if `setProps` isn't set. If we don't set the `value` on `componentWillReceiveProps`, then callbacks wouldn't work anymore.

 We could add a `mode` prop that if set to `auto` it will ignore props coming in from `componentWillReceiveProps` or something, but I'd rather not do that if we can avoid it. 

Any other suggestions?

**Comment by chriddyp** [comment]
Here's the logic where we determine whether or not to pass in `setProps` or not, I recommend digging into that:
https://github.com/plotly/dash-renderer/blob/b1cfc996563bd0b57f9487b15212dd5b782e5b3a/src/components/core/NotifyObservers.react.js#L68-L73

From that logic, I forget if that we're checking if that component is an _output_ or if we're checking if it's an _input_ but I believe that its checking if that component has any properties that are `Inputs`, meaning that they need to share their updates with `dash-renderer` so that it can fire callbacks.

If so, you're right in that the component could be an output, so we still need to check if any props _have changed_, we can't just rely on `this.state`.

And since the `selected` property itself could be an output (another control could update which tab is selected), we need to update our `this.state.selected` if the property has changed.

However, no other property can change within the tab itself (via user interaction), so we don't need to hold anything else in state, we only need to keep track of `selected`.

So, could we have logic like this?
```
constructor() {
    this.state = {selected: this.props.selected || this.props.children[0].props.children.value};
}

componentWillReceiveProps(newProps) {
    if(newProps.selected !== this.props.selected) {
         /*
          *  user updated selected via some callback or 
          * the tab was changed by calling `setProps`
          */
         this.setState({selected: newProps.selected});
    }
}

selectHandler(value) {
    if (this.props.setProps) { this.props.setProps({selected: value}); }
    else { this.setState({selected: value}); }
}

render() {
    const {selected} = this.state;
    const {style, ...} = this.props;
    return ...
}
```

In summary:
- Initialize the state with the selected value or a default
- If its controlled by dash-renderer (`setProps` is defined), then use `setProps` to update the selected value
- Otherwise, if it's uncontrolled, then only udpate the state via `this.setState` and ignore the props coming in (as they haven't changed)
- If value changes during a render, then it means that a callback updated it - hydrate the state with the new value

**Comment by valentijnnieman** [comment]
@chriddyp You're absolutely right - I was thinking of it in the wrong way and didn't realize that we could just skip using this.state if setProps was defined. Was a pretty easy fix in the end, I was just going down the wrong rabbit hole I guess! Thanks for your help. I've pushed a fix to #315, maybe you can review it?
